pkg/doc/jose/jwk.go

```diff
@@ -18,107 +18,220 @@ import (
+	if isSecp256k1(j.Algorithm, j.Kty, j.Crv) {
+		var ecPubKey *ecdsa.PublicKey
+
+		ecPubKey, ok := j.Key.(*ecdsa.PublicKey)
+		if !ok {
+			ecPubKey = &j.Key.(*ecdsa.PrivateKey).PublicKey
+		}
+
+		pubKey := &btcec.PublicKey{
+			Curve: btcec.S256(),
+			X:     ecPubKey.X,
+			Y:     ecPubKey.Y,
+		}
+
+		return pubKey.SerializeCompressed(), nil
 	}
 
-	var jwk JWK
+	switch pubKey := j.Public().Key.(type) {
+	case *ecdsa.PublicKey, *rsa.PublicKey, ed25519.PublicKey:
+		pubKBytes, err := x509.MarshalPKIXPublicKey(pubKey)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read get public key: %w", err)
+		}
 
-	err := convertMapToValue(jwkRaw, &jwk)
-	if err != nil {
-		return nil, false
+		return pubKBytes, nil
+	default:
+		return nil, fmt.Errorf("unsupported public key type in kid '%s'", j.KeyID)
 	}
+}
 
-	return &jwk, true
+// UnmarshalJSON reads a key from its JSON representation.
+func (j *JWK) UnmarshalJSON(jwkBytes []byte) error {
+	var key jsonWebKey
+
+	marshalErr := json.Unmarshal(jwkBytes, &key)
+	if marshalErr != nil {
+		return fmt.Errorf("unable to read JWK: %w", marshalErr)
+	}
+
+	if isSecp256k1(key.Alg, key.Kty, key.Crv) {
+		jwk, err := unmarshalSecp256k1(&key)
+		if err != nil {
+			return fmt.Errorf("unable to read JWK: %w", err)
+		}
+
+		*j = *jwk
+	} else {
+		var joseJWK jose.JSONWebKey
+
+		err := json.Unmarshal(jwkBytes, &joseJWK)
+		if err != nil {
+			return fmt.Errorf("unable to read jose JWK, %w", err)
+		}
+
+		j.JSONWebKey = joseJWK
+	}
+
+	j.Kty = key.Kty
+	j.Crv = key.Crv
+
+	return nil
 }
 
-// jsonWebKey contains subset of json web key json properties
-type jsonWebKey struct {
-	Kty string      `json:"kty,omitempty"`
-	Kid string      `json:"kid,omitempty"`
-	Crv string      `json:"crv,omitempty"`
-	X   *byteBuffer `json:"x,omitempty"`
-	Y   *byteBuffer `json:"y,omitempty"`
+// MarshalJSON serializes the given key to its JSON representation.
+func (j *JWK) MarshalJSON() ([]byte, error) {
+	if isSecp256k1(j.Algorithm, j.Kty, j.Crv) {
+		return marshalSecp256k1(j)
+	}
+
+	return (&j.JSONWebKey).MarshalJSON()
 }
 
-// DecodePublicKey reads a public key from its JSON Web Key representation.
-// TODO : to be part of jose.JWK [Issue#1513]
-func DecodePublicKey(jwkBytes []byte) ([]byte, error) {
-	key := &jsonWebKey{}
+func isSecp256k1(alg, kty, crv string) bool {
+	return strings.EqualFold(alg, secp256k1Alg) ||
+		(strings.EqualFold(kty, secp256k1Kty) && strings.EqualFold(crv, secp256k1Crv))
+}
 
-	err := json.Unmarshal(jwkBytes, key)
-	if err != nil {
-		return nil, fmt.Errorf("unable to read JWK, %w", err)
+func unmarshalSecp256k1(jwk *jsonWebKey) (*JWK, error) {
+	if jwk.X == nil {
+		return nil, fmt.Errorf("missed x value in EC key")
 	}
 
-	if strings.EqualFold(key.Kty, secp256k1Kty) && strings.EqualFold(key.Crv, secp256k1Crv) {
-		//  if kty="EC" and Crv="secp256k1", then handle differently
-		return btcPublicKey(key.X, key.Y)
+	if jwk.Y == nil {
+		return nil, fmt.Errorf("missed y value in EC key")
```

Comment:
I noticed there are many detailed error messages returned like the above line.

If these messages reveal key details, then it's a security vulnerability. Please review all the returned errors in this document and only return generic messages to the user (or probably log detailed error messages in `debug` mode only).

Reply:
I agree. I changed the detailed error messages to a single generic `invalid JWK`.